### PR TITLE
DataSegment memory optimizations

### PR DESCRIPTION
**Pull request by leventov**
- Deduplicated `DataSegment` contents (loadSpec's keys, dimensions and metrics lists as a whole) more aggressively;
 - Used `ArrayMap` instead of default `LinkedHashMap` for DataSegment.loadSpec, because they have only 3 entries on average;
 - Pruned DataSegment.loadSpec on Brokers

### CODE REVIEW

**Comment by leventov** [comment]
Is it expected that we have two almost identical classes called "DataSegmentTest" in different modules?

**Comment by himanshug** [comment]
> Is it expected that we have two almost identical classes called "DataSegmentTest" in different modules?

If they are testing different things and can't be merged into single class, then its fine. Or else, please merge them.

> Pruned DataSegment.loadSpec on Brokers
Can we just drop them from announcements made from historicals/realtime-processes instead ? I think @nishantmonu51 already made a change at some point to optionally drop metric/dimension names at some point.

**Comment by leventov** [comment]
@himanshug could you point to a PR?

**Comment by himanshug** [comment]
https://github.com/druid-io/druid/pull/2784

**Comment by leventov** [comment]
@himanshug thanks. I think it's different because it doesn't affect the presence of the data on coordinator endpoints. And it's a more decoupled approach.

**Comment by himanshug** [comment]
There are endpoints at coordinator ( https://github.com/druid-io/druid/blob/master/server/src/main/java/io/druid/server/http/MetadataResource.java ) to get source of truth details of a Segment information from DB.
So, I'm not sure if there is any benefit in keeping that information even in Coordinator memory.

In fact I would change even the default behavior to not publish dimension, metrics, loadSpecs information in announcements so that all users get that.

does anyone else have an opinion ? @gianm @drcrallen @nishantmonu51

**Comment by himanshug** [comment]
it was decided in dev-sync today to go with current state of this PR wherein historicals announce the loadSpec and they are dropped by broker only.

**Comment by leventov** [comment]
@himanshug @gianm could you please take a look?

**Comment by gianm** [comment]
@leventov i'll take a look now

File: api/src/main/java/io/druid/timeline/DataSegment.java [hunk 1]
```diff
@@ -52,15 +54,13 @@
...
-    @Override
-    public String apply(String input)
-    {
-      return interner.intern(input);
-    }
-  };
+  private static final Interner<String> stringInterner = Interners.newWeakInterner();
+  private static final Interner<List<String>> dimensionsInterner = Interners.newWeakInterner();
+  private static final Interner<List<String>> metricsInterner = Interners.newWeakInterner();
+  private static final Map<String, Object> PRUNED_LOAD_SPEC = ImmutableMap.of(
```
[the remark this answers is not in the document]
Comment (gianm) [reply api/src/main/java/io/druid/timeline/DataSegment.java]: Would also suggest adding a note to the docs for any relevant methods, something like "This method does not return loadSpecs for segments; if you need those then you can use the Coordinator API XXX."

File: api/src/main/java/io/druid/timeline/DataSegment.java [hunk 2]
```diff
@@ -52,15 +55,14 @@
...
-  private static final Interner<String> interner = Interners.newWeakInterner();
-  private static final Function<String, String> internFun = new Function<String, String>()
-  {
-    @Override
-    public String apply(String input)
-    {
-      return interner.intern(input);
-    }
-  };
+  private static final Interner<String> stringInterner = Interners.newWeakInterner();
```
[the remark this answers is not in the document]
Comment (leventov) [reply api/src/main/java/io/druid/timeline/DataSegment.java]: Renamed

File: api/src/main/java/io/druid/timeline/DataSegment.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (leventov) [reply api/src/main/java/io/druid/timeline/DataSegment.java]: No, as far as I can tell. It's exposed only in `MetadataResource` and `DatasourcesResource`, which are present only on Coordinator node.

**Comment by leventov** [comment]
@gianm do you have more comments on this PR?

**Comment by gianm** [comment]
@leventov LGTM; sorry, am a few days behind on github email at the moment.

**Pull request merged**